xla/service/gpu/cudnn_fused_conv_rewriter_test.cc

```diff
@@ -295,6 +319,105 @@ TEST_F(CudnnFusedConvRewriterTest, TestBias) {
+
+      input = TYPE[1,<<<format>>>,64] parameter(0)
+      filter = TYPE[3,3,64,64] parameter(1)
+      bias = TYPE[64] parameter(2)
+
+      conv = TYPE[1,<<<format>>>,64] convolution(input, filter), window={size=3x3 pad=1_1x1_1}, dim_labels=b01f_01io->b01f, feature_group_count=1
+      broadcasted_bias = TYPE[1,<<<format>>>,64] broadcast(bias), dimensions={3}
+      add1 = TYPE[1,<<<format>>>,64] add(conv, broadcasted_bias)
+      ROOT relu = TYPE[1,<<<format>>>,64] maximum(zeros, add1)
+    })";
+    absl::flat_hash_map<absl::string_view, absl::string_view> replacements;
+    replacements["<<<format>>>"] = "3,3";
+    TestMatchWithAllTypes(absl::StrReplaceAll(code, replacements));
+    replacements["<<<format>>>"] = "5,5";
+    TestMatchWithAllTypes(absl::StrReplaceAll(code, replacements));
+    replacements["<<<format>>>"] = "3,3";
+    TestMatchWithAllTypes(absl::StrReplaceAll(code, replacements));
+}
+
+TEST_F(CudnnFusedConvRewriterTest, TestBiasMultiCallMultiThread) {
+  // max(0, conv(x, w) + bias);
+  std::string code = R"(
+    HloModule Test
+
+    ENTRY Test {
+      zero = TYPE[] constant(0)
+      zeros = TYPE[1,<<<format>>>,64] broadcast(zero), dimensions={}
+
+      input = TYPE[1,<<<format>>>,64] parameter(0)
+      filter = TYPE[3,3,64,64] parameter(1)
+      bias = TYPE[64] parameter(2)
+
+      conv = TYPE[1,<<<format>>>,64] convolution(input, filter), window={size=3x3 pad=1_1x1_1}, dim_labels=b01f_01io->b01f, feature_group_count=1
+      broadcasted_bias = TYPE[1,<<<format>>>,64] broadcast(bias), dimensions={3}
+      add1 = TYPE[1,<<<format>>>,64] add(conv, broadcasted_bias)
+      ROOT relu = TYPE[1,<<<format>>>,64] maximum(zeros, add1)
+    })";
+    absl::flat_hash_map<absl::string_view, absl::string_view> replacements;
+    replacements["<<<format>>>"] = "3,3";
+    replacements["TYPE"] = "f16";
+    const std::string hlo = absl::StrReplaceAll(code, replacements);
+    std::atomic<int> counter(-1);
+
+    // this seems very wasteful, but HloModule's are apparently not reusable
+    std::unique_ptr<HloModule> modules[100];
+    for(int i=0; i<100; i++) {
+      TF_ASSERT_OK_AND_ASSIGN(auto module, ParseAndReturnVerifiedModule(hlo));
+      DebugOptions debug_opts = module->config().debug_options();
+      debug_opts.set_xla_gpu_use_runtime_fusion(true);
+      module->mutable_config().set_debug_options(debug_opts);
+      modules[i] = std::move(module);
+    }
+
+    auto executor = [&](int id) {
+      for(int iter=0; iter<25; iter++) {
+        if(id==0)
+          counter++;
+        else if(id==1 && iter<5)
+          while(counter < iter);
+        else if(id==2 && iter==0)
+          while(counter < 0);
+        else if(id==3 && iter==0)
+          while(counter < 5);
+        EXPECT_TRUE(Run(std::move(modules[id*25+iter]), true));
+        fflush(stdout);
+      }
+    };
+    std::thread threads[4];
+    for(int i=0; i<4; i++)
+      threads[i] = std::thread(executor, i);
```

Comment:
What is the motivation for the multi thread test? All threads run the same module, right?

Reply:
This would potentially flag race conditions due to MIOpen (or CuDNN) executing identical fused operations in different threads.

Reply:
The race condition would need to make `Run(...)` return `false` for this test to catch it, which may or may not happen. Wouldn't running under ThreadSanetizer give you better chances of finding such issues? Also, this test class doesn't seem the right place to test MIOpen or cuDNN for thread safety, in my opinion.

Reply:
The test is still there, but it was changed to use RunAndCompare instead of Run, so it would catch mismatches. I don't see any other places where backend implementations of fused convolutions are tested or could be tested for concurrency issues. 

As it is, worst thing that might happen is that there might be an issue but the test won't catch it. I can certainly remove the test, but I'm not sure how dropping the test would improve matters.

Reply:
Tests need to be maintained and in its current form this test will be hard to maintain. When this test breaks, the person looking at the breakage needs to right information to understand the issue.

* A code comment should explain the motivation for this test, because it is different from the non-threaded tests around it.
* Please explain where the race condition would occur. I'm not sure here - the threads don't seem to share anything and don't communicate.
* This test does not test XLA's conv rewriter, but MIOpen / cuDNN, right? This is not obvious and deserves a comment too. Would XLA actually call convs on multiple threads (inter-op parallelism)?

Reply:
The threads don't share anything and they don't communicate, but, at least in the MIOpen case, stream executor will cache and reuse the same "fusion plan" handle for operations with identical parameters (https://github.com/openxla/xla/blob/main/xla/stream_executor/rocm/rocm_dnn.cc#L560), and potentially the same handle may be used in parallel in different threads. CuDNN may be doing something similar at some level.

Since this PR is almost 2 months old by now, rather than argue the point further, I'm going to drop the test. If  race conditions are really possible, someone will have fun tracking them down the hard way.